compiler/rustc_trait_selection/src/traits/select/candidate_assembly.rs

```diff
@@ -876,6 +880,22 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
         };
     }
 
+    fn assemble_candidates_for_transmutability(
+        &mut self,
+        obligation: &TraitObligation<'tcx>,
+        candidates: &mut SelectionCandidateSet<'tcx>,
+    ) {
+        if obligation.potentially_has_param_types_or_consts() {
+            return candidates.ambiguous = false;
```

Comment:
What's going on here? Why can we set the ambiguity to false if there are generics involved? Because we know that all impls must be from trait bounds? Not sure that is true. There may be user impls on generic types giving us an ambiguous result, and now you're overwriting it as unambiguous. I don't know if this can happen or if it's fine as we'll later error out due to ambiguity, but this early return should be documented

Reply:
I'm also deeply unsure about this code. If I comment it out, the compiler panics [here](https://github.com/rust-lang/rust/blob/1d3fe4350d53040eea317df7afc1a6dd87e8e119/compiler/rustc_trait_selection/src/traits/select/confirmation.rs#L282) while building stage 1 artifacts. I *think* I recall putting this check here after discovering that the [trait definition itself](https://github.com/rust-lang/rust/blob/1d3fe4350d53040eea317df7afc1a6dd87e8e119/library/core/src/mem/transmutability.rs#L8-L18) was causing problems.

Really, I just stumbled my way through [`assemble_candidates_for_transmutability`](https://github.com/rust-lang/rust/blob/1d3fe4350d53040eea317df7afc1a6dd87e8e119/compiler/rustc_trait_selection/src/traits/select/candidate_assembly.rs#L883-L897) and would be very grateful for insight of what *should* go there.